### ISSUE TITLE
Always cross compile

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nixpkgs#21471]
Exherbo, another linux distribution that has done good work with cross-compiling, always builds cross compilers---i.e. `--host` `--build` and `--target` are always passed to gcc even if some of those flags are the same: http://git.exherbo.org/arbor.git/tree/packages/sys-devel/gcc/gcc.exlib#n187. The big advantage here is by using the cross-compiling code path in all cases, there's less to maintain, and actual cross-compiling is less likely to rot.

In https://github.com/NixOS/nixpkgs/pull/21268 (specifically https://github.com/NixOS/nixpkgs/pull/21268/commits/633feb4e39a24c7da9ff4ae1b5f4219e0c2a1bb6 but I'll probably rebase at some point breaking that link), I introduce always-defined `buildPlatform` `hostPlatform` and `targetPlatform` for the same reasons (I keep around nullable `crossSystem` and not-always present `stdenv.cross` for compatibility). So if/when that PR is merged would be a good time to tackle this.

Naturally this would be a gcc-caused mass rebuild on Linux. Darwin shouldn't be as affected as LLVM, by default, includes all the targets we care about.

**Comment by Ericson2314** [comment]
@DavidEGrayson's work in https://github.com/DavidEGrayson/nixcrpkgs/ should greatly accelerate this!

**Comment by Ericson2314** [comment]
@cleverca22 points out https://github.com/taktoa/arcane-chat/blob/master/default.nix has a bunch of hacks needed to get windows to work. Would be good to keep these things in mind.

**Comment by Ericson2314** [comment]
OK, hope this actually happens this cycle :crossed_fingers:.

**Comment by wmertens** [comment]
This will be seriously amazing!

So after a quick look, this change requires changing all occurrences of `crossSystem` to `hostPlatform` or `targetPlatform`? Is there no way to guess the correct value at evaluation time?

In other words, it's a huge single-shot change and after that it will just work?

If so, I propose we declare a week "cross-compile week" and only merge changes that make this work. Once everything builds on staging we merge to master and we're done?

Or is there more to it? Will there be things that break?

**Comment by Ericson2314** [comment]
The `crossSystem` vs `hostPlatform` things are actually all taken care of, now! What's left is:

 - [ ] #36867 LLVM-based cross compilation
 - [ ] Change `configureFlags` on `binutils` and `gcc` to be `[ "host" "build" "target" ]`. Watch out for this arm (`armv5` I think it is?) case where then things fail to build. Also do for `ghc` and any other compiler where we have `configureFlags` set today. Also make the various cases of `targetPrefix` unconditionally `hostPlatform.config "-"`. (Must be done with the previous.

**Comment by lheckemann** [comment]
Said alternative is setting the `build_alias`, `host_alias` and `target_alias` environment variables. However, I'm not sure if this is really a desirable solution, both because it'll only work for autotools-based configure scripts and because I'm not sure this way of setting the platforms is actually considered a public/documented/stable interface by autotools.

**Comment by wmertens** [comment]
So when everything is cross-compiled, will the hashes of something
cross-compiled on platform A to platform T be the same as platform B to T?

**Comment by lheckemann** [comment]
No. The derivation hashes will be different, and as a result the output hashes will also be different. In an ideal scenario the only differences between cross-compiled and natively-compiled outputs will be the store paths, but I highly doubt that we'll achieve that in practice.

**Comment by wmertens** [comment]
Well, maybe we can achieve that with a CAS store, but only if the bytecode
is 100% the same… It would be great if we could exclude the host platform
from the input hashes…

**Comment by Ericson2314** [comment]
@wmertens Yes that is a goal, but it will be very hard. CAS / intentional store is necessary, but not sufficient, as the input binaries for the build tools would *not* be anywhere near bit-for-bit the same.

[Also, you mean build platform not host platform, BTW. The names are indeed awful.]

**Comment by lheckemann** [comment]
@dezgeg raised the concern of tests not being run when cross-compiling. Does exherbo maybe have a solution that we can steal?

**Comment by eternaleye** [comment]
@lheckemann: My experience on Exherbo has been that tests run just fine when build == host (build == run), and would likely run just fine any time host (run) is runnable (such as i686 or x32 on a 64-bit builder). That could be further expanded using `binfmt_misc`and qemu-user.

**Comment by Ericson2314** [comment]
I can't find it well on my phone, but the answer must be in https://git.exherbo.org/paludis/paludis.git/

**Comment by eternaleye** [comment]
[This](https://git.exherbo.org/paludis/paludis.git/diff/paludis/repositories/e/ebuild/exheres-0/src_test.bash?h=cross&id=bfab9490c76d6c3f136cb58640f3f713aac2e991) is the relevant bit, but is overconservative - it just [checks for build == host (build == run)](https://git.exherbo.org/paludis/paludis.git/diff/paludis/repositories/e/ebuild/exheres-0/build_functions.bash?h=cross&id=bfab9490c76d6c3f136cb58640f3f713aac2e991) - though it misnames 'run' as 'target'.

**Comment by Ericson2314** [comment]
@Eternaleye err we mean how are (autoconf-based) configure scripts invoked. I see `default_src_configure` but not where the various env vars it uses are defined.

**Comment by eternaleye** [comment]
@Ericson2314: All the `DEFAULT_SRC_TEST_*` variables are set by individual packages.

You probably want to look at the [definition of `econf`](https://git.exherbo.org/paludis/paludis.git/tree/paludis/repositories/e/ebuild/exheres-0/build_functions.bash?h=cross#n126) instead.

**Comment by Ericson2314** [comment]
@eternaleye 
https://git.exherbo.org/paludis/paludis.git/tree/paludis/repositories/e/ebuild/exheres-0/build_functions.bash?h=cross#n155 huh so `--build` and `--host` indeed are always passed if `CBUILD` and `CHOST` are defined. Do you all indeed never have configure run programs then? Brave!

**Comment by eternaleye** [comment]
That's a work in progress. The most problematic cases are actually thing like gobject-introspection, which is inherently incompatible with cross-compilation at the architectural level. Things like that work in practice when build == host (build == run), but fail when "really" cross-compiling. The ability to [ban commands in the environment](https://git.exherbo.org/paludis/paludis.git/commit/?h=cross&id=1725c41781dc7148c6410c9a5197e61ada90d835) is the main tool used to fix such issues where possible - for example, [we ban unprefixed pkg-config](https://git.exherbo.org/arbor.git/tree/packages/dev-util/pkg-config/pkg-config-0.29.2-r2.exheres-0#n123) so we can learn what needs patched to respect `--host`.

**Comment by Ericson2314** [comment]
Neat patch!

**Comment by dezgeg** [comment]
Indeed that patch looks like it solves many of those problems in practice. But my general point stands, aka even if you do "cross compilation" with e.g. both build and host set to x86, you still wouldn't detect those problems, you still would have to do e.g. and x86 -> ARM cross compile to actually detect them, right?

**Comment by eternaleye** [comment]
@degzeg Yes; there's really no feasible way to do static analysis of autotools-based buildsystems.

**Comment by Gaelan** [comment]
I'm running into this when trying to compile for aarch32 on aarch64. I've been sending PRs to set `configurePlatforms = ["host" "build"];` on a per-package basis, but is there anything stopping us from doing it globally at this point?

**Comment by wmertens** [comment]
Let's put it in staging and watch the world burn :)

**Comment by Ericson2314** [comment]
Still pertinent, and in fact https://github.com/NixOS/nixpkgs/pull/87909 is getting new attention.

**Comment by sternenseemann** [comment]
We did make some progress towards this: #123419, #119625, … (?)

**Comment by wmertens** [comment]
One future concern: would we be ok with declaring cross-compiled binaries as equivalent, no matter what the build platform was?

If so, provided we have a trusted mapping (being looked at with the intensional efforts), how would we go about generating all the input hashes that map to the same build?

**Comment by Ericson2314** [comment]
@wmertens I don't think we should calculate them up front. The space of possible build platforms is very, very large.

**Comment by wmertens** [comment]
@Ericson2314 another option would be to somehow replace the build platform input hash with a placeholder when calculating the input hash? Then all cross-compiling build tools for a target platform would map to the same dependency input hash?

**Comment by Lahvuun** [comment]
I'll use this chance to mention that I've been looking into cross compilation with Nix, and my conclusion is that the build platform shouldn't go into an input hash at all. It makes no sense to have two bit-for-bit identical packages with different hashes simply because one was built natively and another one cross compiled. The *host* platform (on which the package will be used) should be the one included in the input hash. And for packages that have no native code (for example text files, like configuration or Python programs) platform shouldn't matter at all.

If you get rid of that, then the other source of difference is the compiler. Here, you basically want native and cross compilers built from the same source with the same codegen options to have different store paths, but affect input hashes in the same way. I don't think this is possible without extending Nix, and even then this kind of rule bending feels like it goes against the core Nix philosophy.

**Comment by Ericson2314** [comment]
@lahvuun I suspect with CA derivations we could do "cache cheating as a plugin" to allow people to experiment with mixing builds without baking in logic in Nix contra to our morals.

**Comment by Mindavi** [comment]
This PR is just about done now: https://github.com/NixOS/nixpkgs/pull/87909

I'm thinking it might be merged just after branch-off of 22.05?

**Comment by sternenseemann** [comment]
Only a step towards it, however, since prefixing the compiler by default is still missing.

**Comment by Mindavi** [comment]
Yeah, that'll be an interesting exercise
